[SPARK-12630][Python][MLlib][DOC] Update param descriptions in classification.py

vijaykiran opened the pull request:
Updates the `param` descriptions in python mllib's classification.py to be consistent. See [SPARK-11219] for more
details.

vijaykiran committed:
[SPARK-12630][DOC] Update param descriptions

Updates the `param` descriptions consistent. See [SPARK-11219] for more
details.

AmplabJenkins commented:
Can one of the admins verify this patch?

BryanCutler commented on python/pyspark/mllib/classification.py:
```diff
@@ -323,38 +330,45 @@ def train(cls, data, iterations=100, initialWeights=None, regParam=0.01, regType
+          (default: 0.01).
+        :param regType:
+          The type of regularizer used for training our model.
+
```
remove blank lines here, and around other "Allowed values"

BryanCutler commented on python/pyspark/mllib/classification.py:
```diff
@@ -490,37 +506,45 @@ def train(cls, data, iterations=100, step=1.0, regParam=0.01,
+            - None for no regularization
+
+          (default: "l2")
+
```
remove blank lines

BryanCutler commented on python/pyspark/mllib/classification.py:
```diff
@@ -272,37 +274,42 @@ def train(cls, data, iterations=100, step=1.0, miniBatchFraction=1.0,
+        :param intercept:
+          Boolean parameter which indicates the use or not of the augmented
+          representation for training data (i.e. whether bias features are
+          activated or not).
```
Line length can go up to 100 characters, can you please collapse these lines and others to that limit?  There might have been a different limit before, but no reason to use so many lines now.

BryanCutler commented on python/pyspark/mllib/classification.py:
```diff
@@ -323,38 +330,45 @@ def train(cls, data, iterations=100, initialWeights=None, regParam=0.01, regType
+          activated or not)
+         (default: False)
+        :param corrections:
+           The number of corrections used in the LBFGS update.
```
there is an extra space indentation here

BryanCutler commented on python/pyspark/mllib/classification.py:
```diff
@@ -323,38 +330,45 @@ def train(cls, data, iterations=100, initialWeights=None, regParam=0.01, regType
+          Boolean parameter which indicates the use or not of the augmented
+          representation for training data (i.e. whether bias features are
+          activated or not)
+         (default: False)
```
this is not aligned properly

BryanCutler commented:
thanks @vijaykiran!  I marked a few things for correction and I think in general we should extend the comments to the 100 character limit where applicable.

BryanCutler commented on python/pyspark/mllib/classification.py:
```diff
@@ -323,38 +330,45 @@ def train(cls, data, iterations=100, initialWeights=None, regParam=0.01, regType
+          The training data, an RDD of LabeledPoint.
+        :param iterations:
+          The number of iterations
+          (default: 100).
```
nit: no period after the ending parenthesis in default line

BryanCutler commented on python/pyspark/mllib/classification.py:
```diff
@@ -94,16 +94,18 @@ class LogisticRegressionModel(LinearClassificationModel):
+      Regression. In Multinomial Logistic Regression, the intercepts will not
+      be a single value, so the intercepts will be part of the weights.)
+    :param numFeatures:
+      the dimension of the features.
```
nit:  capitalize the first word in the description sentence

vijaykiran committed:
Style fixes

- Style fixes based on review comments by @BryanCutler.
- Changed fill-column to 100 instead of 80.

vijaykiran commented:
@BryanCutler  Thanks for the review, I added a new commit - will review update other two PRs with 100 fill-column as well.

BryanCutler commented:
LGTM.  It's not necessary to squash commits, but could you update the PR title/description to indicate that this is for PySpark MLlib classification?

ping @jkbradley @mengxr

jkbradley commented:
I just added a note to the parent JIRA about a formatting issue affecting all 5 PRs: [https://issues.apache.org/jira/browse/SPARK-11219?focusedCommentId=15090225&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15090225]
Could you please check it out & ping when I should review again?  Thank you!

BryanCutler commented:
Hi @vijaykiran , will you be able to update the param descriptions to be only up to 74 character limit?  Sorry for the mixup, but I think this should be mergable after that.

vijaykiran committed:
Limit parameter desciptions to 74 columns

vijaykiran commented:
ping @BryanCutler - Updated the param descriptions, can you take a look ?

BryanCutler commented:
I viewed in pydoc with a window size of 80 and LGTM, cc @jkbradley to take a look.  Thanks @vijaykiran!

BryanCutler commented on python/pyspark/mllib/classification.py:
```diff
@@ -91,19 +91,21 @@ def predict(self, test):
-            classification problem in Multinomial Logistic Regression.
-            By default, it is binary logistic regression so numClasses
-            will be set to 2.
+    Classification model trained using Multinomial/Binary Logistic Regression.
```
This line is over the 74 char limit

BryanCutler commented:
@vijaykiran , I missed one line that is a bit too long, could you please fix that, otherwise LGTM

vijaykiran committed:
A couple of more 74 column fixes

vijaykiran commented:
ping @BryanCutler @jkbradley - can you guys please take a look again?

BryanCutler commented:
LGTM

mengxr commented:
@vijaykiran @BryanCutler Please run `make html` under `python/docs` and check the warning messages. This is what I got:

~~~
/Users/meng/src/spark/python/pyspark/mllib/classification.py:docstring of pyspark.mllib.classification.LogisticRegressionWithSGD.train:23: ERROR: Unexpected indentation.
/Users/meng/src/spark/python/pyspark/mllib/classification.py:docstring of pyspark.mllib.classification.LogisticRegressionWithSGD.train:26: WARNING: Block quote ends without a blank line; unexpected unindent.
/Users/meng/src/spark/python/pyspark/mllib/classification.py:docstring of pyspark.mllib.classification.LogisticRegressionWithLBFGS.train:17: ERROR: Unexpected indentation.
/Users/meng/src/spark/python/pyspark/mllib/classification.py:docstring of pyspark.mllib.classification.LogisticRegressionWithLBFGS.train:20: WARNING: Block quote ends without a blank line; unexpected unindent.
/Users/meng/src/spark/python/pyspark/mllib/classification.py:docstring of pyspark.mllib.classification.SVMWithSGD.train:23: ERROR: Unexpected indentation.
/Users/meng/src/spark/python/pyspark/mllib/classification.py:docstring of pyspark.mllib.classification.SVMWithSGD.train:26: WARNING: Block quote ends without a blank line; unexpected unindent.
/Users/meng/src/spark/python/pyspark/mllib/regression.py:docstring of pyspark.mllib.regression.LinearRegressionWithSGD:3: ERROR: Unexpected indentation.
/Users/meng/src/spark/python/pyspark/mllib/regression.py:docstring of pyspark.mllib.regression.LinearRegressionWithSGD:4: WARNING: Block quote ends without a blank line; unexpected unindent.
/Users/meng/src/spark/python/pyspark/mllib/regression.py:docstring of pyspark.mllib.regression.RidgeRegressionWithSGD:3: ERROR: Unexpected indentation.
/Users/meng/src/spark/python/pyspark/mllib/regression.py:docstring of pyspark.mllib.regression.RidgeRegressionWithSGD:4: WARNING: Block quote ends without a blank line; unexpected unindent.
/Users/meng/src/spark/python/pyspark/mllib/regression.py:docstring of pyspark.mllib.regression.LassoWithSGD:3: ERROR: Unexpected indentation.
/Users/meng/src/spark/python/pyspark/mllib/regression.py:docstring of pyspark.mllib.regression.LassoWithSGD:4: WARNING: Block quote ends without a blank line; unexpected unindent.
/Users/meng/src/spark/python/pyspark/mllib/regression.py:docstring of pyspark.mllib.regression.IsotonicRegression:7: ERROR: Unexpected indentation.
/Users/meng/src/spark/python/pyspark/mllib/regression.py:docstring of pyspark.mllib.regression.IsotonicRegression:12: ERROR: Unexpected indentation.
~~~

vijaykiran commented:
@mengxr thanks, will take a look today

mengxr commented:
@vijaykiran Thanks! Please ignore the warnings from `regression.py`. That will be addressed in https://issues.apache.org/jira/browse/SPARK-12986.

mengxr commented:
ok to test

SparkQA commented:
**[Test build #50152 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50152/consoleFull)** for PR 10598 at commit [`bf8f8a0`](https://github.com/apache/spark/commit/bf8f8a0b82b49b9144449cd00872345dba3bf061).

SparkQA commented:
**[Test build #50152 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50152/consoleFull)** for PR 10598 at commit [`bf8f8a0`](https://github.com/apache/spark/commit/bf8f8a0b82b49b9144449cd00872345dba3bf061).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50152/
Test PASSed.

mengxr commented:
@vijaykiran Could you fix the doc warnings in `classification.py`? Thanks!

BryanCutler commented on python/pyspark/mllib/classification.py:
```diff
@@ -272,37 +275,42 @@ def train(cls, data, iterations=100, step=1.0, miniBatchFraction=1.0,
+            - "l1" for using L1 regularization
+            - "l2" for using L2 regularization
+            - None for no regularization
+          (default: "l2")
```
It looks like we have a problem with the formatting of this param.  If we want the html to look the same and get rid of any warnings, then I think it has to be written like this
```
:param regType:
  The type of regularizer used for training our model.
  
  :Allowed values:
    - "l1" for using L1 regularization
    - "l2" for using L2 regularization
    - None for no regularization

  (default: "l2")
```
which, I think, looks ugly in the code and not even that great in the html.  Alternatively, we could rewrite it inline similar to how some other params have it, like this
```
:param regType:
  The type of regularizer used for training our model.
  Allowed values: "l1" for using L1 regularization;
  "l2" for using L2 regularization; None for no
  regularization.
  (default: "l2")
```
Not perfect, but both code and html look decent, imho.
@mengxr , would you be ok with the second inline format?

mengxr replied:
The first one looks better but no strong preference. I tried the following that also works:

~~~
        :param regType:
          The type of regularizer used for training our model.
          Allowed values:

            - "l1" for using L1 regularization
            - "l2" for using L2 regularization (default)
            - None for no regularization
~~~

Note `Allow values` is not a Sphinx keyword.

BryanCutler replied:
It's not a huge deal so I'm ok with either, but my issue with the additional blank lines is that in the context of the already large param list, the extra blank lines seem like a little much and the rendered html looks sloppy to me too.  Since I noticed other parts describing allowed values in a single paragraph, I thought maybe we should stick to that.

mengxr replied:
If we use `:Allowed values:`, it shows up as `:Allowed values:` in html, which is weird. We should change existing ones instead of repeating the mistakes, in a separate PR. I don't think the rendered html changes due to that additional blank line. It is just control by top margin of a list, which could be configured via CSS. This is how it looks in html for my proposal, which looks good to me:

![screen shot 2016-02-11 at 2 21 41 pm](https://cloud.githubusercontent.com/assets/829644/12992699/cb339390-d0ca-11e5-980f-779b0b97c97a.png)

Let's quickly fix this one and merge this PR.

BryanCutler replied:
I believe that `:Allowed values:` would need to be preceded by a blank line for Sphinx to interpret it in bold like before, but I don't think that needs to be done here.  What @mengxr proposed above looks fine with me too.

@vijaykiran , will you be able to make this fix soon?  I could take over and finish this last little bit if you can't.

BryanCutler commented:
Thanks for working on this @vijaykiran , I finished up the remaining fix in https://github.com/apache/spark/pull/11183 so please close this one up.
cc @mengxr

mengxr commented:
@vijaykiran I merged #11183 into master. Do you mind closing this PR? Thanks!
